elm/wizard.py

```diff
@@ -311,3 +238,310 @@ def chat(self, query,
+                                        aws_session_token=session_token)
+
+        super().__init__(model, token_budget=token_budget)
+
+    def get_embedding(self, text):
+        """Get the 1D array (list) embedding of a text string
+        as generated by AWS Titan.
+
+        Parameters
+        ----------
+        text : str
+            Text to embed
+
+        Returns
+        -------
+        embedding : list
+            List of float that represents the numerical embedding of the text
+        """
+
+        body = json.dumps({"inputText": text, })
+
+        model_id = self.EMBEDDING_MODEL
+        accept = 'application/json'
+        content_type = 'application/json'
+
+        response = self._aws_client.invoke_model(
+            body=body,
+            modelId=model_id,
+            accept=accept,
+            contentType=content_type
+        )
+
+        response_body = json.loads(response['body'].read())
+        embedding = response_body.get('embedding')
+
+        return embedding
+
+    def query_vector_db(self, query, limit=100):
+        """Returns a list of strings and relatednesses, sorted from most
+        related to least.
+
+        Parameters
+        ----------
+        query : str
+            Question being asked of GPT
+        limit : int
+            Number of top results to return.
+
+        Returns
+        -------
+        strings : np.ndarray
+            1D array of related strings
+        score : np.ndarray
+            1D array of float scores of strings
+        ids : np.ndarray
+            1D array of IDs in the text corpus corresponding to the
+            ranked strings/scores outputs.
+        """
+
+        query_embedding = self.get_embedding(query)
+
+        self.cursor.execute("SELECT ewiz_kb.id, "
+                            "ewiz_kb.chunks, "
+                            "ewiz_kb.embedding <=> %s::vector as score "
+                            "FROM ewiz_schema.ewiz_kb "
+                            "ORDER BY embedding <=> %s::vector LIMIT %s;",
+                            (query_embedding, query_embedding, limit,), )
+
+        result = self.cursor.fetchall()
+
+        strings = [s[1] for s in result]
+        scores = [s[2] for s in result]
+        best = [s[0] for s in result]
+
+        return strings, scores, best
+
+    def make_ref_list(self, ids):
+        """Make a reference list
+
+        Parameters
+        ----------
+        used_index : np.ndarray
+            IDs of the used text from the text corpus
+
+        Returns
+        -------
+        ref_list : list
+            A list of references (strs) used. Ideally, this is something like:
+            ["{ref_title} ({ref_url})"]
+        """
+
+        placeholders = ', '.join(['%s'] * len(ids))
+
+        sql_query = ("SELECT ewiz_kb.title, ewiz_kb.url "
+                     "FROM ewiz_schema.ewiz_kb "
+                     "WHERE ewiz_kb.id IN (" + placeholders + ")")
```

Comment:
Same comment here about `ewiz_schema` and `ewiz_kb`. they should be modifyiable by the user. Also, it seems like it should be really easy to make a new table `ewiz_meta` with `id` as the primary key that we can query here.